packages/compass-preferences-model/src/preferences.ts

```diff
@@ -503,16 +556,68 @@ class Preferences {
    * @returns The currently active set of preferences.
    */
   getPreferences(): GlobalPreferences {
+    return this._computePreferenceValuesAndStates().values;
+  }
+
+  private _getStoredValues(): GlobalPreferences {
     return {
       ...this._userPreferencesModel.getAttributes({
         props: true,
         derived: true,
       }),
       ...this._globalPreferences.cli,
       ...this._globalPreferences.global,
+      ...this._globalPreferences.hardcoded,
     };
   }
 
+  private _computePreferenceValuesAndStates() {
+    const values = this._getStoredValues();
+    const states: Partial<Record<string, PreferenceState>> = {};
+    for (const key of Object.keys(this._globalPreferences.cli))
+      states[key] = 'set-cli';
+    for (const key of Object.keys(this._globalPreferences.global))
+      states[key] = 'set-global';
+    for (const key of Object.keys(this._globalPreferences.hardcoded))
+      states[key] = 'hardcoded';
+
+    const originalValues = { ...values };
+    const originalStates = { ...states };
+
+    function deriveValue<K extends keyof GlobalPreferences>(
+      key: K
+    ): {
+      value: GlobalPreferences[K];
+      state: PreferenceState;
+    } {
+      const descriptor = allPreferencesProps[key];
+      if (!descriptor.deriveValue) {
+        return { value: originalValues[key], state: originalStates[key] };
+      }
+      return (
+        descriptor.deriveValue as DeriveValueFunction<GlobalPreferences[K]>
+      )(
+        // `as unknown` to work around TS bug(?) https://twitter.com/addaleax/status/1572191664252551169
+        (k) =>
+          (k as unknown) === key ? originalValues[k] : deriveValue(k).value,
+        (k) =>
+          (k as unknown) === key ? originalStates[k] : deriveValue(k).state
+      );
+    }
```

Comment:
@alenakhineika I don’t think we specified in the tech design _how_ `networkTraffic` interacts with the individual preference values. I chose to use a generic mechanism for this rather than hardcoding the preference names and their interactions, also considering that for `enableShell` and `readOnly` we are going to have to do something similar again.